app/helpers/locations_helper.rb

```diff
@@ -42,12 +42,43 @@ def display_location_attributes(location)
     base_attributes
   end
 
+  ActivePeriod = Struct.new(:from, :to)
+  def active_periods(location)
+    periods = [ActivePeriod.new(location.available_from, nil)]
+
+    sorted_deactivation_periods = remove_nested_periods(location.location_deactivation_periods.sort_by(&:deactivation_date))
+    sorted_deactivation_periods.each do |deactivation|
+      periods.find { |period| period.to.nil? }.to = deactivation.deactivation_date
+      periods << ActivePeriod.new(deactivation.reactivation_date, nil)
+    end
+
+    remove_overlapping_and_empty_periods(periods)
+  end
+
   def location_availability(location)
-    availability = "Active from #{location.available_from.to_formatted_s(:govuk_date)}"
-    location.location_deactivation_periods.each do |deactivation|
-      availability << " to #{(deactivation.deactivation_date - 1.day).to_formatted_s(:govuk_date)}\nDeactivated on #{deactivation.deactivation_date.to_formatted_s(:govuk_date)}"
-      availability << "\nActive from #{deactivation.reactivation_date.to_formatted_s(:govuk_date)}" if deactivation.reactivation_date.present?
+    availability = ""
+    active_periods(location).each do |period|
+      if period.from.present?
+        availability << "\nActive from #{period.from.to_formatted_s(:govuk_date)}"
+        availability << " to #{(period.to - 1.day).to_formatted_s(:govuk_date)}\nDeactivated on #{period.to.to_formatted_s(:govuk_date)}" if period.to.present?
+      end
     end
-    availability
+    availability.strip
+  end
+
+private
+
+  def remove_overlapping_and_empty_periods(periods)
+    periods.select { |period| period.from.present? && (period.to.nil? || period.from < period.to) }
```

Comment:
```suggestion
    periods.select { |(from, to)| from.present? && (to.nil? || from < to) }
```

Part 2 if you use my suggestion without structs. Again: not sure if this is actually better.